engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java

```diff
@@ -0,0 +1,258 @@
+    final DirectBuffer variables = MsgPackUtil.asMsgPack(activated.getValue().getVariables());
+    final Record<JobRecordValue> completedRecord =
+        engineRule.job().withVariables(variables).completeAndWait(activated.getKey());
+
+    // then
+    Assertions.assertThat(completedRecord.getMetadata())
+        .hasRecordType(RecordType.EVENT)
+        .hasIntent(JobIntent.COMPLETED);
+    MsgPackUtil.assertEquality(
+        MsgPackHelper.EMTPY_OBJECT, completedRecord.getValue().getVariables());
+  }
+
+  @Test
+  public void shouldThrowExceptionOnCompletionIfVariablesAreInvalid() {
+    // given
+    engineRule.createJob(JOB_TYPE, PROCESS_ID);
+    final Record<JobBatchRecordValue> batchRecord =
+        engineRule.jobs().withType(JOB_TYPE).activateAndWait();
+
+    final byte[] invalidVariables = new byte[] {1}; // positive fixnum, i.e. no object
+
+    // when
+    final Throwable throwable =
+        catchThrowable(
+            () ->
+                engineRule
+                    .job()
+                    .withVariables(new UnsafeBuffer(invalidVariables))
+                    .complete(batchRecord.getValue().getJobKeys().get(0)));
+
+    // then
+    assertThat(throwable).isInstanceOf(RuntimeException.class);
+    assertThat(throwable.getMessage()).contains("Property 'variables' is invalid");
+    assertThat(throwable.getMessage())
+        .contains("Expected document to be a root level object, but was 'INTEGER'");
+  }
+
+  @Test
+  public void shouldRejectCompletionIfJobIsCompleted() {
+    // given
+    engineRule.createJob(JOB_TYPE, PROCESS_ID);
+    final Record<JobBatchRecordValue> batchRecord =
+        engineRule.jobs().withType(JOB_TYPE).activateAndWait();
+
+    final DirectBuffer variables =
+        MsgPackUtil.asMsgPack(batchRecord.getValue().getJobs().get(0).getVariables());
+    final Long jobKey = batchRecord.getValue().getJobKeys().get(0);
+    engineRule.job().withVariables(variables).complete(jobKey);
+
+    // when
+    engineRule.job().withVariables(variables).complete(jobKey);
+    final Record<JobRecordValue> rejection =
+        RecordingExporter.jobRecords(JobIntent.COMPLETE)
+            .withRecordType(RecordType.COMMAND_REJECTION)
+            .getFirst();
+
+    // then
+    Assertions.assertThat(rejection.getMetadata())
+        .hasRejectionType(RejectionType.NOT_FOUND)
+        .hasIntent(JobIntent.COMPLETE);
+  }
+
+  @Test
+  public void shouldRejectCompletionIfJobIsFailed() {
+    // given
+    engineRule.createJob(JOB_TYPE, PROCESS_ID);
+
+    // when
+    final Record<JobBatchRecordValue> batchRecord =
+        engineRule.jobs().withType(JOB_TYPE).activateAndWait();
+    final Long jobKey = batchRecord.getValue().getJobKeys().get(0);
+    engineRule.job().fail(jobKey);
+
+    engineRule
+        .job()
+        .withVariables(
+            MsgPackUtil.asMsgPack(batchRecord.getValue().getJobs().get(0).getVariables()))
+        .complete(jobKey);
+    final Record<JobRecordValue> rejection =
```

Comment:
Please take a look at the `PublishMessageClient` or at other clients how I implement the expected rejection